packages/tiled/lib/src/common/property.dart

```diff
@@ -12,35 +12,213 @@ part of tiled;
+    final type = parser.getPropertyType('type', defaults: PropertyType.string);
+
+    switch (type) {
+      case PropertyType.object:
+        return ObjectProperty(
+          name: name,
+          value: parser.getInt('value', defaults: 0),
+        );
+
+      case PropertyType.color:
+        return ColorProperty(
+          name: name,
+          value: parser.getColor('value', defaults: const Color(0x00000000)),
+          hexValue: parser.getString('value', defaults: '#00000000'),
+        );
+
+      case PropertyType.bool:
+        return BoolProperty(
+          name: name,
+          value: parser.getBool('value', defaults: false),
+        );
+
+      case PropertyType.float:
+        return FloatProperty(
+          name: name,
+          value: parser.getDouble('value', defaults: 0),
+        );
+
+      case PropertyType.int:
+        return IntProperty(
+          name: name,
+          value: parser.getInt('value', defaults: 0),
+        );
+
+      case PropertyType.file:
+        return FileProperty(
+          name: name,
+          value: parser.getString('value', defaults: '.'),
+        );
+
+      case PropertyType.string:
+        final value = parser.formatSpecificParsing((json) {
+          return json.getString('value', defaults: '');
+        }, (xml) {
+          final attrString = parser.getStringOrNull('value');
+          if (attrString != null) {
+            return attrString;
+          } else {
+            // In tmx files, multi-line text property values can be stored
+            // inside the <property> node itself instead of in the 'value'
+            // attribute
+            return xml.element.innerText;
+          }
+        });
+
+        return StringProperty(
+          name: name,
+          value: value,
         );
+    }
+  }
+}
+
+/// A wrapper for a Tiled property set
+///
+/// Accessing an int value
+/// ```dart
+/// properties.get<int>('foo') == 3
+/// ```
+///
+/// Accessing an int property:
+/// ```dart
+/// properties.getProp<IntProperty>('foo') ==
+///     IntProperty(name: 'foo', value: 3);
+/// ```
+class CustomProperties extends Iterable<Property> {
+  static const empty = CustomProperties({});
+
+  /// The properties, indexed by name
+  final Map<String, Property> byName;
+
+  const CustomProperties(this.byName);
+
+  /// Get a property value by its name.
+  ///
+  /// [T] must be the type of the value, which depends on the property type.
+  /// The following Tiled properties map to the follow Dart types:
+  ///  - int property -> int
+  ///  - float property -> double
+  ///  - boolean property -> bool
+  ///  - string property -> string
+  ///  - color property -> ui.Color
+  ///  - file property -> string (path)
+  ///  - object property -> int (ID)
+  T? getValue<T>(String name) {
+    return getProperty(name)?.value as T;
+  }
+
+  /// Get a typed property by its name
+  T? getProperty<T extends Property<dynamic>>(String name) {
+    return byName[name]! as T;
```

Comment:
These methods don't need a nullable return type since they are both forced to an object and then cast to T. Or the return should be written in a way that can return null. I wonder why the analyzer didn't pick up on this. 🤔

Reply:
I meant to remove the `!`, they should be nullable in reality, right? It's map of unknown things and the caller may not know if this particular property was set on the tile/obj/etc

Reply:
Just removing the `!` is not enough though since you'd cast the value to a non-nullable type so you'll have to change to `as T?` too.

Reply:
👍  Fixed